File: work/stellar-restomod/find_lights.py
import json
import struct

with open('stellar_setup.glb', 'rb') as f:
    header = f.read(20)
    magic, version, length, json_length, json_format = struct.unpack('<IIIII', header)
    json_data = f.read(json_length)

data = json.loads(json_data.decode('utf-8'))

print("--- MATERIALS ---")
if 'materials' in data:
    for idx, mat in enumerate(data['materials']):
        name = mat.get('name', '')
        emissive = mat.get('emissiveFactor', [0,0,0])
        print(f"Material {idx}: Name='{name}' Emissive={emissive}")

print("\n--- NODES WITH MESHES ---")
node_map = {}
if 'nodes' in data:
    for idx, node in enumerate(data['nodes']):
        node_map[idx] = node

def get_absolute_translation(node_idx):
    t = [0.0, 0.0, 0.0]
    curr_idx = node_idx
    visited = set()
    while curr_idx is not None and curr_idx not in visited:
        visited.add(curr_idx)
        node = node_map.get(curr_idx)
        if not node:
            break
        if 'translation' in node:
            t[0] += node['translation'][0]
            t[1] += node['translation'][1]
            t[2] += node['translation'][2]
        
        # Find parent
        parent = None
        for i, n in node_map.items():
            if 'children' in n and curr_idx in n['children']:
                parent = i
                break
        curr_idx = parent
    return t

if 'nodes' in data:
    for node_idx, node in node_map.items():
        if 'mesh' in node:
            mesh_idx = node['mesh']
            mesh = data['meshes'][mesh_idx]
            mesh_name = mesh.get('name', '')
            abs_pos = get_absolute_translation(node_idx)
            
            primitives = mesh.get('primitives', [])
            mat_indices = [p.get('material') for p in primitives if p.get('material') is not None]
            mat_names = []
            for m_idx in mat_indices:
                if 'materials' in data and m_idx < len(data['materials']):
                    mat_names.append(data['materials'][m_idx].get('name', ''))
                else:
                    mat_names.append('default')
            
            print(f"Node {node_idx}: Name='{node.get('name', '')}', Mesh {mesh_idx}: Name='{mesh_name}', Pos=[{', '.join(f'{v:.3f}' for v in abs_pos)}], Materials={mat_names}")

File: work/stellar-restomod/find_lights.js
const fs = require('fs');

const buffer = fs.readFileSync('stellar_setup.glb');
const jsonLength = buffer.readUInt32LE(12);
const jsonBuffer = buffer.slice(20, 20 + jsonLength);
const jsonStr = jsonBuffer.toString('utf-8');
const data = JSON.parse(jsonStr);

console.log('--- MATERIALS ---');
if (data.materials) {
    data.materials.forEach((mat, idx) => {
        console.log(`Material ${idx}: Name='${mat.name || ""}' Emissive=${JSON.stringify(mat.emissiveFactor || [0,0,0])}`);
    });
}

console.log('\n--- NODES WITH MESHES ---');
const nodeMap = new Map();
if (data.nodes) {
    data.nodes.forEach((node, idx) => {
        nodeMap.set(idx, node);
    });
}

function getAbsoluteTranslation(nodeIdx) {
    let t = [0, 0, 0];
    let currIdx = nodeIdx;
    let visited = new Set();
    while (currIdx !== undefined && currIdx !== null && !visited.has(currIdx)) {
        visited.add(currIdx);
        const node = nodeMap.get(currIdx);
        if (!node) break;
        if (node.translation) {
            t[0] += node.translation[0];
            t[1] += node.translation[1];
            t[2] += node.translation[2];
        }
        // Find parent
        let parent = null;
        if (data.nodes) {
            for (let i = 0; i < data.nodes.length; i++) {
                if (data.nodes[i].children && data.nodes[i].children.includes(currIdx)) {
                    parent = i;
                    break;
                }
            }
        }
        currIdx = parent;
    }
    return t;
}

if (data.nodes) {
    data.nodes.forEach((node, nodeIdx) => {
        if (node.mesh !== undefined) {
            const mesh = data.meshes[node.mesh];
            const meshName = mesh.name || "";
            const absPos = getAbsoluteTranslation(nodeIdx);
            
            // Check materials used in primitives
            const primitives = mesh.primitives || [];
            const matIndices = primitives.map(p => p.material).filter(m => m !== undefined);
            const matNames = matIndices.map(idx => data.materials[idx] ? data.materials[idx].name || "" : "default");

            console.log(`Node ${nodeIdx}: Name='${node.name || ""}', Mesh ${node.mesh}: Name='${meshName}', Pos=[${absPos.map(v => v.toFixed(3)).join(', ')}], Materials=[${matNames.join(', ')}]`);
        }
    });
}
